events/eventservices/eventsService.go

```diff
@@ -0,0 +1,221 @@
+package eventservices
+
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/FactomProject/factomd/common/constants/runstate"
+	"github.com/FactomProject/factomd/common/globals"
+	"github.com/FactomProject/factomd/common/interfaces"
+	"github.com/FactomProject/factomd/events"
+	"github.com/FactomProject/factomd/events/eventmessages"
+	"github.com/FactomProject/factomd/events/eventoutputformat"
+	"github.com/FactomProject/factomd/p2p"
+	"github.com/FactomProject/factomd/util"
+	"github.com/gogo/protobuf/proto"
+	log "github.com/sirupsen/logrus"
+	"net"
+	"time"
+)
+
+var eventService events.EventService
+var eventServiceControl events.EventServiceControl
+
+const (
+	defaultProtocol           = "tcp"
+	defaultConnectionHost     = "127.0.0.1"
+	defaultConnectionPort     = 8040
+	defaultOutputFormat       = eventoutputformat.Protobuf
+	sendRetries               = 3
+	dialRetryPostponeDuration = 5 * time.Minute
+	redialSleepDuration       = 10 * time.Second
+)
+
+type eventServiceInstance struct {
+	params               *EventServiceParams
+	eventsOutQueue       chan *eventmessages.FactomEvent
+	postponeSendingUntil time.Time
+	connection           net.Conn
+	owningState          interfaces.IState
+}
+
+func NewEventService(state interfaces.IState, config *util.FactomdConfig, factomParams *globals.FactomParams) (events.EventService, events.EventServiceControl) {
+	return NewEventServiceTo(state, selectParameters(factomParams, config))
+}
+
+func NewEventServiceTo(state interfaces.IState, params *EventServiceParams) (events.EventService, events.EventServiceControl) {
+	if eventService == nil {
+		eventServiceInstance := &eventServiceInstance{
+			eventsOutQueue: make(chan *eventmessages.FactomEvent, p2p.StandardChannelSize),
+			params:         params,
+			owningState:    state,
+		}
+		eventService = eventServiceInstance
+		eventServiceControl = eventServiceInstance
+		go eventServiceInstance.processEventsChannel()
+	}
+	return eventService, eventServiceControl
+}
+
+func (esi *eventServiceInstance) Send(event events.EventInput) error {
+	if esi.owningState.GetRunState() > runstate.Running { // Stop queuing messages to the events channel when shutting down
+		return nil
+	}
+
+	// Only send info messages when MuteEventsDuringStartup is enabled
+	if esi.params.MuteEventsDuringStartup && !esi.owningState.IsRunLeader() &&
+		event.GetEventSource() != eventmessages.EventSource_NODE_MESSAGE && event.GetEventSource() != eventmessages.EventSource_PROCESS_MESSAGE {
+		return nil
+	}
+
+	factomEvent, err := MapToFactomEvent(event)
+	if err != nil {
+		return fmt.Errorf("failed to map to factom event: %v\n", err)
+	}
+	factomEvent.IdentityChainID = &eventmessages.Hash{
+		HashValue: esi.owningState.GetIdentityChainID().Bytes(),
+	}
+	select {
+	case esi.eventsOutQueue <- factomEvent:
```

Comment:
This is dropping the most recent messages if the queue is full, so the queue grows stale. If we could drop off of the front when the queue is full, that would be ideal. The implementation get's a bit tricky since different threads can add/remove from the queue, so I think we might just have to live with this.

I'd like to suggest adding a counter here that counts the number of messages dropped from back pressure. We could make this counter accessible via a prometheus counter, and via a `/v1/livefeed` endpoint for livefeed specific api calls. This would let a client identify if they are potentially losing information.